crypto/rsa/rsa_sp800_56b_check.c

```diff
@@ -403,6 +403,11 @@ int ossl_rsa_sp800_56b_check_keypair(const RSA *rsa, const BIGNUM *efixed,
         ERR_raise(ERR_LIB_RSA, RSA_R_INVALID_KEYPAIR);
         return 0;
     }
+    /* (Step 3.c): check that the modulus length is a positive even integer */
+    if (nbits <= 0 || (nbits & 0x1)) {
```

Comment:
The only invocation of `ossl_rsa_sp800_56b_check_keypair()` I could find (other than in tests in `test/rsa_sp800_56b_test.c`) is in `ossl_rsa_validate_pairwise()` in `crypto/rsa/rsa_chk.c`, and that already differs based on `#ifdef FIPS_MODULE` and will only call this function when inside the FIPS module, so I don't think we need the additional ifdef here.

Reply:
Then the whole function probably should be in `#ifdef FIPS_MODULE` as it is otherwise just an unused dead code.

Reply:
> Then the whole function probably should be in `#ifdef FIPS_MODULE` as it is otherwise just an unused dead code

So does the PR need to be updated to do this? As OP says, the only call outside of tests is

```
int ossl_rsa_validate_pairwise(const RSA *key)
{
#ifdef FIPS_MODULE
    return ossl_rsa_sp800_56b_check_keypair(key, NULL, -1, RSA_bits(key));
#else
    return rsa_validate_keypair_multiprime(key, NULL) > 0;
#endif
}
```

Reply:
The code is used in the test.

I tried changing the build instructions for the test such that it will use the FIPS-compiled version and ifdef the entire function, but that isn't as simple as I would have hoped it would be:

```diff
diff --git a/test/build.info b/test/build.info
index 73f8de1f7a..0695fdfc95 100644
--- a/test/build.info
+++ b/test/build.info
@@ -909,7 +909,8 @@ IF[{- !$disabled{tests} -}]

     SOURCE[rsa_sp800_56b_test]=rsa_sp800_56b_test.c
     INCLUDE[rsa_sp800_56b_test]=.. ../include ../crypto/rsa ../apps/include
-    DEPEND[rsa_sp800_56b_test]=../libcrypto.a libtestutil.a
+    DEPEND[rsa_sp800_56b_test]=../providers/libfips.a libtestutil.a
+    DEFINE[rsa_sp800_56b_test]=FIPS_MODULE

     IF[{- !$disabled{'deprecated-3.0'} -}]
       PROGRAMS{noinst}=rsa_x931_test
```

makes the test compile, but it crashes on startup when calling `CRYPTO_zalloc` from `CRYPTO_THREAD_lock_new` because the `c_CRYPTO_zalloc` pointer in the FIPS provider is `NULL` until `OSSL_provider_init_int` is called.

I'd appreciate pointers on how to run the test against code compiled for the FIPS provider without opening an entire can of worms.

Reply:
@t8m ^ ?  Otherwise are you okay not to put `ossl_rsa_sp800_56b_check_keypair()` inside `#ifdef FIPS_MODULE`?

Reply:
Hmm... OK. This is for a different PR I suppose if anybody wants to persuade it.